src/run/handlers/tags-handler.cts

```diff
@@ -0,0 +1,146 @@
+  if (tags.length === 0 || !timestamp) {
+    return Promise.resolve(false)
+  }
+
+  const cacheStore = getMemoizedKeyValueStoreBackedByRegionalBlobStore({ consistency: 'strong' })
+
+  //    Full-route cache and fetch caches share a lot of tags
+  //    but we will only do actual blob read once withing a single request due to cacheStore
+  //    memoization.
+  //    Additionally, we will resolve the promise as soon as we find first
+  //    stale tag, so that we don't wait for all of them to resolve (but keep all
+  //    running in case future `CacheHandler.get` calls would be able to use results).
+  //    "Worst case" scenario is none of tag was invalidated in which case we need to wait
+  //    for all blob store checks to finish before we can be certain that no tag is stale.
+  return new Promise<boolean>((resolve, reject) => {
+    const tagManifestPromises: Promise<boolean>[] = []
+
+    for (const tag of tags) {
+      const lastRevalidationTimestampPromise = lastTagRevalidationTimestamp(tag, cacheStore)
+
+      tagManifestPromises.push(
+        lastRevalidationTimestampPromise.then((lastRevalidationTimestamp) => {
+          if (!lastRevalidationTimestamp) {
+            // tag was never revalidated
+            return false
+          }
+          const isStale = lastRevalidationTimestamp >= timestamp
+          if (isStale) {
+            // resolve outer promise immediately if any of the tags is stale
+            resolve(true)
+            return true
+          }
+          return false
+        }),
+      )
+    }
+
+    // make sure we resolve promise after all blobs are checked (if we didn't resolve as stale yet)
+    Promise.all(tagManifestPromises)
+      .then((tagManifestAreStale) => {
+        resolve(tagManifestAreStale.some((tagIsStale) => tagIsStale))
+      })
+      .catch(reject)
+  })
+}
+
+/**
+ * Transform a tag or tags into an array of tags and handle white space splitting and encoding
+ */
+function getCacheTagsFromTagOrTags(tagOrTags: string | string[]): string[] {
+  return (Array.isArray(tagOrTags) ? tagOrTags : [tagOrTags])
+    .flatMap((tag) => tag.split(/,|%2c/gi))
+    .filter(Boolean)
+}
+
+export function purgeEdgeCache(tagOrTags: string | string[]): void {
+  const tags = getCacheTagsFromTagOrTags(tagOrTags)
+
+  if (tags.length === 0) {
+    return
+  }
+
+  const purgeCachePromise = purgeCache({ tags, userAgent: purgeCacheUserAgent }).catch((error) => {
+    // TODO: add reporting here
+    getLogger()
+      .withError(error)
+      .error(`[NetlifyCacheHandler]: Purging the cache for tags [${tags.join(',')}] failed`)
+  })
+
+  getRequestContext()?.trackBackgroundWork(purgeCachePromise)
+}
+
+async function doRevalidateTag(tags: string[]): Promise<void> {
+  getLogger().withFields({ tags }).debug('NetlifyCacheHandler.revalidateTag')
+
+  if (tags.length === 0) {
+    return
+  }
+
+  const data: TagManifest = {
```

Comment:
Maybe `initialTagData`?

Reply:
that's not just initial - this is the only way we are setting when tag was revalidated last - so we do this on first tag invalidation but also on every subsequent one (also this was just copied as is)

maybe it should just be named `tagManifest` instead to be very verbose what it represents because `data` is just ... not telling what it is exactly

Reply:
I know it was simply copied over, but I was searching for a slightly better name. Happy with a rename to `tagManifest` as well.